python/ray/experimental/workflow/workflow_mananger.py

```diff
@@ -0,0 +1,180 @@
+import functools
+from typing import List, Tuple, Union, Any, Dict, Callable, Optional
+import uuid
+
+import ray
+
+from ray.experimental.workflow import workflow_context
+
+RRef = ray.ObjectRef  # Alias ObjectRef because it is too long in type hints
+WorkflowOutputType = Union[RRef, Tuple[RRef]]
+StepArgType = Union[RRef, Any]
+RayTaskOutputType = Union[RRef, List[RRef], Tuple[RRef]]
+
+
+def resolve_object_ref(ref: RRef) -> Tuple[Any, RRef]:
```

Comment:
I feel this is not consistent with the API. Right now we don't take care of nested objects. 
Here we continue deref until we get the physical data, which I don't think we should support. Think of it as nested objects.

For example, 
```
@workflow.step
def return_data() -> Workflow[ObjectRef]:
    obj = ray.put(create_big_data())
    obj2 = ray.put(obj)
    return obj2

@workflow.step
def receive_data(data: ObjectRef):
    obj = ray.get(data)  # OK
    obj2 = ray.get(obj) # NOT OK, right now.

@workflow.step
def entry():
    obj = return_data.step()
    receive_data.step(obj)
```
Keeps derefs makes it hard to figure out what's happening here.
The type of input of a type is becoming hard to figure out later when we have nested structure support.

Reply:
All of this is to avoid using `ray.get` which is incompatible with checkpointing (we discussed a lot about it in early stage docs). Also I have to consider nested objects because they are generated by nested workflow steps, and the fault tolerance design is totally depends on them.